File: behavioral-design-pattern/memento-pattern/memento.go
// Memento Pattern
// ================
// Category: Behavioral Design Pattern
//
// Intent:
//   Without violating encapsulation, capture and externalize an object's
//   internal state so that the object can be restored to this state later.
//
// When to use:
//   - You need to save and restore snapshots of an object's state
//   - A direct interface to obtaining the state would expose implementation
//     details and break encapsulation
//   - You need undo/rollback functionality
//
// Key Participants:
//   - Originator (GameCharacter): the object whose state needs saving;
//     creates a memento containing a snapshot of its current internal state
//   - Memento (GameSave): stores the originator's internal state;
//     protects against access by objects other than the originator
//   - Caretaker (SaveManager): responsible for keeping the memento safe;
//     never operates on or examines the contents of a memento

package main

import (
	"fmt"
	"strings"
	"time"
)

// --- Memento ---

// GameSave is the memento: a snapshot of the character's state.
// Fields are unexported (lowercase) to protect encapsulation in a real project.
// Here they are exported for demonstration clarity.
type GameSave struct {
	Health    int
	Level     int
	PositionX float64
	PositionY float64
	Inventory []string // snapshot copy
	Timestamp string
	Label     string
}

func (s *GameSave) String() string {
	return fmt.Sprintf("[%s] '%s' -- HP:%d Lv:%d Pos:(%.0f,%.0f) Items:[%s]",
		s.Timestamp, s.Label, s.Health, s.Level,
		s.PositionX, s.PositionY, strings.Join(s.Inventory, ", "))
}

// --- Originator ---

// GameCharacter is the originator whose state we want to save and restore.
type GameCharacter struct {
	Name      string
	Health    int
	Level     int
	PositionX float64
	PositionY float64
	Inventory []string
}

func NewGameCharacter(name string) *GameCharacter {
	return &GameCharacter{
		Name:      name,
		Health:    100,
		Level:     1,
		Inventory: []string{},
	}
}

func (c *GameCharacter) TakeDamage(amount int) {
	c.Health -= amount
	if c.Health < 0 {
		c.Health = 0
	}
	fmt.Printf("  %s took %d damage. HP: %d\n", c.Name, amount, c.Health)
}

func (c *GameCharacter) LevelUp() {
	c.Level++
	c.Health = 100 // full heal on level up
	fmt.Printf("  %s leveled up to Lv.%d! HP restored.\n", c.Name, c.Level)
}

func (c *GameCharacter) MoveTo(x, y float64) {
	c.PositionX = x
	c.PositionY = y
	fmt.Printf("  %s moved to (%.0f, %.0f)\n", c.Name, x, y)
}

func (c *GameCharacter) PickUp(item string) {
	c.Inventory = append(c.Inventory, item)
	fmt.Printf("  %s picked up '%s'\n", c.Name, item)
}

// Save creates a memento capturing the current state.
func (c *GameCharacter) Save(label string) *GameSave {
	// Make a snapshot copy of the inventory slice
	invCopy := make([]string, len(c.Inventory))
	copy(invCopy, c.Inventory)

	memento := &GameSave{
		Health:    c.Health,
		Level:     c.Level,
		PositionX: c.PositionX,
		PositionY: c.PositionY,
		Inventory: invCopy,
		Timestamp: time.Now().Format("2006-01-02 15:04:05"),
		Label:     label,
	}
	fmt.Printf("  >> State saved: '%s'\n", label)
	return memento
}

// Restore restores state from a memento.
func (c *GameCharacter) Restore(memento *GameSave) {
	c.Health = memento.Health
	c.Level = memento.Level
	c.PositionX = memento.PositionX
	c.PositionY = memento.PositionY

	// Restore inventory from the snapshot
	c.Inventory = make([]string, len(memento.Inventory))
	copy(c.Inventory, memento.Inventory)

	fmt.Printf("  >> State restored from '%s'\n", memento.Label)
}

func (c *GameCharacter) Status() string {
	return fmt.Sprintf("  %s: HP=%d Lv=%d Pos=(%.0f,%.0f) Inventory=[%s]",
		c.Name, c.Health, c.Level, c.PositionX, c.PositionY,
		strings.Join(c.Inventory, ", "))
}

// --- Caretaker ---

// SaveManager is the caretaker: it stores saves without inspecting their contents.
type SaveManager struct {
	saves map[string]*GameSave
	order []string // preserve insertion order
}

func NewSaveManager() *SaveManager {
	return &SaveManager{
		saves: make(map[string]*GameSave),
		order: []string{},
	}
}

func (m *SaveManager) Store(key string, save *GameSave) {
	if _, exists := m.saves[key]; !exists {
		m.order = append(m.order, key)
	}
	m.saves[key] = save
}

func (m *SaveManager) Load(key string) *GameSave {
	save, ok := m.saves[key]
	if !ok {
		fmt.Printf("  >> No save found for slot '%s'\n", key)
		return nil
	}
	return save
}

func (m *SaveManager) ListSaves() {
	if len(m.saves) == 0 {
		fmt.Println("  No saves stored.")
		return
	}
	fmt.Println("  Stored saves:")
	for _, key := range m.order {
		if save, ok := m.saves[key]; ok {
			fmt.Printf("    [%s] %s\n", key, save)
		}
	}
}

// --- Main ---

func main() {
	hero := NewGameCharacter("Warrior")
	manager := NewSaveManager()

	fmt.Println("=== Starting the adventure ===")
	fmt.Println(hero.Status())
	fmt.Println()

	// Play through some actions
	hero.MoveTo(10, 20)
	hero.PickUp("Iron Sword")
	hero.LevelUp()
	fmt.Println(hero.Status())
	fmt.Println()

	// Save at checkpoint 1
	checkpoint1 := hero.Save("Before the dungeon")
	manager.Store("checkpoint1", checkpoint1)
	fmt.Println()

	// Enter the dungeon -- things go badly
	hero.MoveTo(50, 80)
	hero.TakeDamage(60)
	hero.PickUp("Rusty Key")
	hero.TakeDamage(30)
	fmt.Println(hero.Status())
	fmt.Println()

	// Save the risky state too
	checkpoint2 := hero.Save("Deep in dungeon (low HP)")
	manager.Store("checkpoint2", checkpoint2)
	_ = checkpoint2 // used via manager
	fmt.Println()

	// Even worse...
	hero.TakeDamage(50)
	fmt.Println(hero.Status())
	fmt.Println()

	// List all saves
	fmt.Println("=== Reviewing saves ===")
	manager.ListSaves()
	fmt.Println()

	// Restore to the safe checkpoint
	fmt.Println("=== Restoring to checkpoint 1 ===")
	if save := manager.Load("checkpoint1"); save != nil {
		hero.Restore(save)
	}
	fmt.Println(hero.Status())
}

File: behavioral-design-pattern/memento-pattern/memento.rs
// Memento Pattern
// ================
// Category: Behavioral Design Pattern
//
// Intent:
//   Without violating encapsulation, capture and externalize an object's
//   internal state so that the object can be restored to this state later.
//
// When to use:
//   - You need to save and restore snapshots of an object's state
//   - A direct interface to obtaining the state would expose implementation
//     details and break encapsulation
//   - You need undo/rollback functionality
//
// Key Participants:
//   - Originator (GameCharacter): the object whose state needs saving;
//     creates a memento containing a snapshot of its current internal state
//   - Memento (GameSave): stores the originator's internal state;
//     protects against access by objects other than the originator
//   - Caretaker (SaveManager): responsible for keeping the memento safe;
//     never operates on or examines the contents of a memento

use std::collections::HashMap;
use std::fmt;

// --- Memento ---

/// GameSave is the memento: an immutable snapshot of the character's state.
/// All fields are private to the module; only GameCharacter can create one.
#[derive(Debug, Clone)]
struct GameSave {
    health: i32,
    level: u32,
    position_x: f64,
    position_y: f64,
    inventory: Vec<String>, // owned snapshot
    timestamp: String,
    label: String,
}

impl fmt::Display for GameSave {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        write!(
            f,
            "[{}] '{}' -- HP:{} Lv:{} Pos:({},{}) Items:{:?}",
            self.timestamp, self.label, self.health, self.level,
            self.position_x, self.position_y, self.inventory
        )
    }
}

// --- Originator ---

/// GameCharacter is the originator whose state we want to save and restore.
struct GameCharacter {
    name: String,
    health: i32,
    level: u32,
    position_x: f64,
    position_y: f64,
    inventory: Vec<String>,
}

impl GameCharacter {
    fn new(name: &str) -> Self {
        Self {
            name: name.to_string(),
            health: 100,
            level: 1,
            position_x: 0.0,
            position_y: 0.0,
            inventory: Vec::new(),
        }
    }

    fn take_damage(&mut self, amount: i32) {
        self.health = (self.health - amount).max(0);
        println!("  {} took {} damage. HP: {}", self.name, amount, self.health);
    }

    fn level_up(&mut self) {
        self.level += 1;
        self.health = 100; // full heal on level up
        println!("  {} leveled up to Lv.{}! HP restored.", self.name, self.level);
    }

    fn move_to(&mut self, x: f64, y: f64) {
        self.position_x = x;
        self.position_y = y;
        println!("  {} moved to ({}, {})", self.name, x, y);
    }

    fn pick_up(&mut self, item: &str) {
        self.inventory.push(item.to_string());
        println!("  {} picked up '{}'", self.name, item);
    }

    // --- Memento creation and restoration ---

    /// Create a memento capturing the current state.
    fn save(&self, label: &str) -> GameSave {
        let memento = GameSave {
            health: self.health,
            level: self.level,
            position_x: self.position_x,
            position_y: self.position_y,
            inventory: self.inventory.clone(), // snapshot copy
            timestamp: chrono_placeholder(),
            label: label.to_string(),
        };
        println!("  >> State saved: '{}'", label);
        memento
    }

    /// Restore state from a memento.
    fn restore(&mut self, memento: &GameSave) {
        self.health = memento.health;
        self.level = memento.level;
        self.position_x = memento.position_x;
        self.position_y = memento.position_y;
        self.inventory = memento.inventory.clone(); // restore from snapshot
        println!("  >> State restored from '{}'", memento.label);
    }

    fn status(&self) -> String {
        format!(
            "  {}: HP={} Lv={} Pos=({},{}) Inventory={:?}",
            self.name, self.health, self.level,
            self.position_x, self.position_y, self.inventory
        )
    }
}

/// Simple timestamp placeholder (avoids external crate dependency).
fn chrono_placeholder() -> String {
    // In production, use the `chrono` crate for real timestamps.
    "2026-04-07 12:00:00".to_string()
}

// --- Caretaker ---

/// SaveManager is the caretaker: it stores mementos without inspecting them.
struct SaveManager {
    saves: HashMap<String, GameSave>,
    /// Insertion order tracking (HashMap doesn't preserve order)
    order: Vec<String>,
}

impl SaveManager {
    fn new() -> Self {
        Self {
            saves: HashMap::new(),
            order: Vec::new(),
        }
    }

    fn store(&mut self, key: &str, save: GameSave) {
        if !self.saves.contains_key(key) {
            self.order.push(key.to_string());
        }
        self.saves.insert(key.to_string(), save);
    }

    fn load(&self, key: &str) -> Option<&GameSave> {
        let save = self.saves.get(key);
        if save.is_none() {
            println!("  >> No save found for slot '{}'", key);
        }
        save
    }

    fn list_saves(&self) {
        if self.saves.is_empty() {
            println!("  No saves stored.");
            return;
        }
        println!("  Stored saves:");
        for key in &self.order {
            if let Some(save) = self.saves.get(key) {
                println!("    [{}] {}", key, save);
            }
        }
    }
}

// --- Main ---

fn main() {
    let mut hero = GameCharacter::new("Warrior");
    let mut manager = SaveManager::new();

    println!("=== Starting the adventure ===");
    println!("{}", hero.status());
    println!();

    // Play through some actions
    hero.move_to(10.0, 20.0);
    hero.pick_up("Iron Sword");
    hero.level_up();
    println!("{}", hero.status());
    println!();

    // Save at checkpoint 1
    let checkpoint1 = hero.save("Before the dungeon");
    manager.store("checkpoint1", checkpoint1);
    println!();

    // Enter the dungeon -- things go badly
    hero.move_to(50.0, 80.0);
    hero.take_damage(60);
    hero.pick_up("Rusty Key");
    hero.take_damage(30);
    println!("{}", hero.status());
    println!();

    // Save the risky state too
    let checkpoint2 = hero.save("Deep in dungeon (low HP)");
    manager.store("checkpoint2", checkpoint2);
    println!();

    // Even worse...
    hero.take_damage(50);
    println!("{}", hero.status());
    println!();

    // List all saves
    println!("=== Reviewing saves ===");
    manager.list_saves();
    println!();

    // Restore to the safe checkpoint
    println!("=== Restoring to checkpoint 1 ===");
    if let Some(save) = manager.load("checkpoint1") {
        hero.restore(save);
    }
    println!("{}", hero.status());
}
